File: C_Sharp/Framework/LazyLoad/LazyLoadDomainObject.cs
using Framework.Data_Manipulation;
using Framework.Domain;

namespace Framework.LazyLoad
{
    public interface ILazyLoadDomainObject<TSearchInput>
    {
        TSearchInput Criteria { get; }
        void Load<TEntity>(ILazyLoader<TEntity, TSearchInput> loader, TSearchInput criteria)
           where TEntity : LazyLoadDomainObject<TSearchInput>;
    }

    public abstract class LazyLoadDomainObject<TSearchInput> : DomainObject, ILazyLoadDomainObject<TSearchInput>
    {
        private TSearchInput _criteria;

        public TSearchInput Criteria { get { return _criteria; } }

        public LazyLoadDomainObject(
            TSearchInput criteria,
            IBaseMapper mapper
            ) : base(mapper)
        {
            _criteria = criteria;
        }

        public void Load<TEntity>(ILazyLoader<TEntity, TSearchInput> loader, TSearchInput criteria)
            where TEntity : LazyLoadDomainObject<TSearchInput>
        {
            if ((criteria == null) || (loader == null))
                return;

            TEntity currentEntity = (TEntity) this;
            RetrieveMatchingEntityDelegate<TEntity> retrieveMatchingEntity = () =>
            {
                return loader.GetEntity(criteria);
            };

            loader.LoadAllFields(ref currentEntity, retrieveMatchingEntity);

            _criteria = default(TSearchInput);      //Clear criteria
        }
    }
}

File: Android/framework/src/main/java/com/LazyLoad/LazyLoadDomainObject.java
package com.LazyLoad;

import com.DataManipulation.BaseMapperInterfaces.IBaseMapper;
import com.Domain.DomainObject;
import com.Interfaces.ILazyLoadDomainObject;
import com.Interfaces.ILazyLoader;

/**
 * Created by aiko on 4/2/17.
 */

public abstract class LazyLoadDomainObject<TSearchInput> extends DomainObject implements ILazyLoadDomainObject<TSearchInput> {
    TSearchInput _criteria;

    public TSearchInput GetCriteria (){return _criteria;}

    public LazyLoadDomainObject(TSearchInput criteria, IBaseMapper mapper) {
        super(mapper);

        _criteria = criteria;
    }

    public <TEntity extends  LazyLoadDomainObject<TSearchInput>> void Load(ILazyLoader<TEntity, TSearchInput> loader)
    {
        if ((_criteria == null) || (loader == null))
            return;

        TEntity currentEntity = (TEntity) this;

        loader.LoadAllFields(currentEntity, _criteria);

        _criteria = null;
    }
}
